proxy/proxy/client.go

```diff
@@ -292,33 +240,126 @@ func isExpired(cfg *tls.Config) bool {
 	return time.Now().After(cfg.Certificates[0].Leaf.NotAfter)
 }
 
-func (c *Client) cachedCfg(instance string) (string, *tls.Config, string) {
+// startRefresh kicks off a refreshCfg asynchronously, returning a channel that is closed
+// when it's complete and updates the cacheEntry when complete. This function should only
+// be called from the scope of "cachedCfg", which controls the logic around throttling refreshes.
+func (c *Client) startRefresh(instance string, refreshCfgBuffer time.Duration) chan bool {
+	done := make(chan bool)
+	go func() {
+		defer close(done)
+		addr, cfg, ver, err := c.refreshCfg(instance)
+
+		c.cacheL.Lock()
+		old := c.cfgCache[instance]
+		// if we failed to refresh cfg do not throw out potentially valid one
+		if err != nil && !isExpired(old.cfg) {
+			logging.Errorf("failed to refresh the ephemeral certificate for %s, returning previous cert instead: %v", instance, err)
+			addr, cfg, ver, err = old.addr, old.cfg, old.version, old.err
+		}
+		c.cfgCache[instance] = cacheEntry{
+			lastRefreshed: time.Now(),
+			err:           err,
+			addr:          addr,
+			version:       ver,
+			cfg:           cfg,
+			done:          done,
+		}
+		c.cacheL.Unlock()
+
+		certExpiration := cfg.Certificates[0].Leaf.NotAfter
+		now := time.Now()
+		timeToRefresh := certExpiration.Sub(now) - refreshCfgBuffer
+		if timeToRefresh <= 0 {
+			// If a new certificate expires before our buffer has expired,
+			// we should wait a bit and schedule a new refresh to much closer to the expiration's date
+			// TODO: This situation probably only occurs for Issue #622 - when the oauth2 token isn't refreshed before the cert is
+			timeToRefresh = certExpiration.Sub(now) - (5 * time.Second)
+			logging.Errorf("new ephemeral certificate expires sooner than expected (adjusting refresh time to compensate): current time: %v, certificate expires: %v", now, certExpiration)
+		}
+		go c.refreshCertAfter(instance, timeToRefresh)
+	}()
+	return done
+}
+
+// isValid returns true if the cacheEntry is still useable
+func isValid(c cacheEntry) bool {
```

Comment:
I generally prefer sticking with positive checks - go/tott-530

Reply:
Interesting. TIL.

Reply:
Actually, reading the link above, I see it says, "If your conditional looks like “if not … else …” then reverse it to put the positive case first.".

That makes me think we flip the method to `isInvalid`, or to stick with a positive phrasing, `isStale`, or `isErred`, etc. That way we write `if isStale` instead of requiring readers to parse `if !isValid`, which requires a little more mental effort to understand.

Reply:
We're not using an `if not … else …`, we're just using an "if else", so that doesn't apply here. If we were, it would be better to put the `isValid` case first and then `!isValid` case second. 

The problem is that the "In" from "Invalid" is a negation, and if you ever need to inverse a negation you end up having a double negative to accomplish that (`!IsInvalid`). 

IMO, the summary wraps it up quite nicely:
>  in general checking the presence or absence of a positive is simpler for readers to understand than checking the presence or absence of a negative.